src/lib/utils/DY.Lib.SplitFunction.fst

```diff
@@ -0,0 +1,287 @@
+  reveal_opaque (`%mk_global_fun) (mk_global_fun);
+  introduce
+    forall tagged_data.
+      match func.decode_tagged_data tagged_data with
+      | Some (tag, raw_data) ->
+        tag `func.tag_belong_to` tag_set ==> (func.apply_global_fun (mk_global_fun func tagged_local_funs) tagged_data == func.apply_local_fun local_fun raw_data)
+      | None -> True
+  with (
+    match func.decode_tagged_data tagged_data with
+    | Some (tag, raw_data) -> (
+      if tag `func.tag_belong_to` tag_set then (
+        mk_global_fun_correct_aux func tagged_local_funs tag_set local_fun tag;
+        func.apply_mk_global_fun (mk_global_fun_aux func tagged_local_funs) tagged_data
+      ) else ()
+    )
+    | None -> ()
+  )
+
+/// Equivalence theorem for `mk_global_fun`.
+/// This may be useful to lift properties of the local functions
+/// to property on the global function.
+/// (e.g. the function keep the same output when the trace grows.)
+
+val mk_global_fun_eq:
+  func:split_function_input_values -> tagged_local_funs:list (func.tag_set_t & func.local_fun_t) ->
+  tagged_data:func.tagged_data_t ->
+  Lemma (
+    func.apply_global_fun (mk_global_fun func tagged_local_funs) tagged_data == (
+      match func.decode_tagged_data tagged_data with
+      | Some (tag, raw_data) -> (
+        match find_local_fun func tagged_local_funs tag with
+        | Some tagged_local_fun -> func.apply_local_fun tagged_local_fun raw_data
+        | None -> func.apply_global_fun func.default_global_fun tagged_data
+      )
+      | None -> func.apply_global_fun func.default_global_fun tagged_data
+    )
+  )
+let mk_global_fun_eq func tagged_local_funs tagged_data =
+  reveal_opaque (`%mk_global_fun) (mk_global_fun);
+  func.apply_mk_global_fun (mk_global_fun_aux func tagged_local_funs) tagged_data
+
+/// If a global function contains some local function,
+/// and the global function input decodes to a tag for this local function,
+/// then the global function is equivalent to the local function on this input.
+
+val local_eq_global_lemma:
+  func:split_function_input_values ->
+  global_fun:func.global_fun_t -> tag_set:func.tag_set_t -> local_fun:func.local_fun_t ->
+  tagged_data:func.tagged_data_t -> tag:func.tag_t -> raw_data:func.raw_data_t ->
+  Lemma
+  (requires
+    func.decode_tagged_data tagged_data == Some (tag, raw_data) /\
+    tag `func.tag_belong_to` tag_set /\
+    has_local_fun func global_fun (tag_set, local_fun)
+  )
+  (ensures func.apply_global_fun global_fun tagged_data == func.apply_local_fun local_fun raw_data)
+let local_eq_global_lemma func global_fun tag_set tagged_local_fun tagged_data tag raw_data = ()
+
+/// In the common case where tag disjointness is unequality,
+/// `all_disjoint` is equivalent to the `no_repeats_p` predicate
+/// from F*'s standard library.
+
+val unequal:
+  #a:Type ->
+  a -> a -> prop
+let unequal #a x y = x =!= y
+
+val no_repeats_p_implies_all_disjoint:
```

Comment:
i think this name should contain unequal, since it is only true for "all_disjoint unequal"

Reply:
Done